src/python/pants/backend/codegen/protobuf/python/python_protobuf_module_mapper.py

```diff
@@ -37,12 +39,15 @@ async def map_protobuf_to_python_modules(
         for tgt in protobuf_targets
     )
 
-    modules_to_addresses: Dict[str, Tuple[Address]] = {}
-    modules_with_multiple_owners: Set[str] = set()
+    # NB: There should be only one address per module, else it's ambiguous.
+    modules_to_addresses: dict[str, tuple[Address]] = {}
+    modules_with_multiple_owners: DefaultDict[str, set[Address]] = defaultdict(set)
```

Comment:
Would it be possible to push all ambiguity detection out into `module_mapper` in order to reduce redundancy between rule implementations? Perhaps by returning the mappings in `FirstPartyPythonMappingImpl` as tuples rather than a dict?

Reply:
No, the definition of ambiguous depends on the implementation. The built in impl allows for the same module to be used by a .py and .pyi file, for example. That cannot be factored up.

Reply:
But couldn't it return an entire non-mapping mapping from `module -> target`, with multiple entries per module? Then the consuming code would group by module and report all of the targets?

Reply:
> But couldn't it return an entire non-mapping mapping from module -> target, with multiple entries per module? Then the consuming code would group by module and report all of the targets?

This is already what is happening. Each specific implementation returns a `FrozenDict[str, tuple[Address, ...]]`. There can be multiple entries per module if that makes sense within the context of that implementation, e.g. the handling of `.pyi` files. But within the vast majority of implementations, it is not valid to have >1 target per module, and it's ambiguous.

> Then the consuming code would group by module and report all of the targets?

This does happen, although we assume that if the same module is covered by >1 implementation, it must be ambiguous. There's no use case yet to allow a module to have multiple owners, each coming from a different implementation.

Is this what you had meant?

Reply:
I'll re-read this after lunch, sorry. But if each implementation was generating a flat list rather than needing to detect duplicates, I would not expect `python_protobuf_module_mapper.py` to need to detect which files have multiple owners (into a variable `modules_with_multiple_owners`)... they could all just be appended to a list.

Reply:
The main takeaway for why that cannot happen is that the validity of duplicates is very context-specific. Duplicates are never legal for most `FirstPartyPythonMappingImpl`s, including Protobuf, but they are legal in a particular case with the core `FirstPartyPythonMappingImpl` due to type stubs. Because it is context-specific, we cannot factor up some generic handling of duplicates.